microsoft/botbuilder-js
# [WIP] Add functional test package

## Description
In this pull request, we created the `functional-tests` package for managing all the test that hit services. We created a new NPM script which will run all the test including the functional tests.
Also, we started using environment variables instead of local variables. 

## Specific Changes
- Create a new `functional-test` package
- Add script for `functional-test` in root package.json
- Move the test that hit services to a functional test project
- Migrate environment variables
- Remove the fusebox reference in the prep-test.cmd file

- Removed the .env file
  *Note:* Currently, the tests that use the variables of the .env file are the ones of _connector test_. Right now those tests are skipped. Due to this file was deleted, if you want to run those test the following variables must be placed as Pipeline variables:

  CLIENT_ID
  CLIENT_SECRET
  HOST_URL=https://slack.botframework.com
  USER_ID=U3Z9ZUDK5:T03CWQ0QB 
  BOT_ID=B21S8SG7J:T03CWQ0QB
	
  Notice that the CLIENT_ID and CLIENT_SECRET do not have values in the .env file. 

- Add the functional-test column Badge in the Readme

## Steps to configure a Pipeline
These are the steps for configuring an Azure Pipeline that run the Functional Tests.

Once this PR is merged, the "Nightly-Build" Pipeline must be configured the following way:

1. Create a new Pipeline and set the NPM custom command to run the `functional-test` script.
![image](https://user-images.githubusercontent.com/37461749/56689408-61f1ec80-66b1-11e9-9c38-9354c89975fb.png)

	
2. Configure the following variables as Pipeline Variables:
	
    From luisRecognizer.test.js we migrate
    LUISAPPID
    
    From luisSdk.test.js we migrate
    LUISSDKAPPID, LUISSDKENDPOINTKEY, LUISENDPOINT
		     
    The figure below shows the Pipeline variables once there were configured and marked as Secret.
    ![image](https://user-images.githubusercontent.com/37461749/56689448-7afa9d80-66b1-11e9-87f4-757c7c92609a.png)

    Note: You can mark every variable as Secret on every variable's field, as you can see in the figure. 
    ![image](https://user-images.githubusercontent.com/37461749/56691661-e3984900-66b6-11e9-8a8a-76969480b5ac.png)


3. Once the variables were configured, you have to add a PowerShell Script Task to take those variables and pass them as environment variables:	
![image](https://user-images.githubusercontent.com/37461749/56689508-9cf42000-66b1-11e9-9a85-ba22b7a12c61.png)

    This is the PowerShell script used. 
    ```
    echo '##vso[task.setvariable variable=LUISAPPID]$(LUISAPPIDSECRET)'
    echo '##vso[task.setvariable variable=LUISSDKAPPID]$(LUISSDKAPPIDSECRET)'
    echo '##vso[task.setvariable variable=LUISSDKENDPOINTKEY]$(LUISSDKENDPOINTKEYSECRET)'
    ```

## Steps to add the functional-test badge information.

1. Click in the 'Status badge' option, as shown in the picture below:	
![image](https://user-images.githubusercontent.com/37461749/56685839-379c3100-66a9-11e9-8bef-bf62aa5fdb2d.png)

2. Copy the Markdown Link, as you can see in the following screen.	
![image](https://user-images.githubusercontent.com/37461749/56689606-e17fbb80-66b1-11e9-93e1-c5b1b88164ac.png)

3. Add the Markdown link in the last part of the version tag in which you want to display the functional-test badge.

    Note: Separate the new content with a vertical bar character '|'
    ![image](https://user-images.githubusercontent.com/37461749/56693120-09731d00-66ba-11e9-97a5-101c5f5ad710.png)
